base/sysinfo.jl

```diff
@@ -321,4 +323,82 @@ windows_version
 
 const WINDOWS_VISTA_VER = v"6.0"
 
+"""
+    Sys.isexecutable(path::String)
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function isexecutable(path::String)
+    @static if iswindows()
+        return isfile(path)
+    else
+        # We use `access()` and `X_OK` to determine if a given path is
+        # executable by the current user.  `X_OK` comes from `unistd.h`.
+        X_OK = 0x01
+        ccall(:access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
+    end
+end
+isexecutable(path::AbstractString) = isexecutable(String(path))
+
+"""
+    Sys.which(program_name::String)
+
+Given a program name, searches the current `PATH` to find the first binary with
+the proper executable permissions that can be run, and returns the absolute
+path. Raises `ArgumentError` if no such program is available.  If a relative or
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::String)
+    # Build a list of program names that we're going to try
+    program_names = String[]
+    @static if iswindows()
+        # If the file already has an extension, try that name first
+        if !isempty(splitext(program_name)[2])
+            push!(program_names, program_name)
+        end
+
+        # But also try appending .exe and .com`
+        for pe in (".exe", ".com")
+            push!(program_names, string(program_name, pe))
+        end
+    else
+        # On non-windows, we just always search for what we've been given
+        push!(program_names, program_name)
+    end
+
+    path_dirs = String[]
+    program_dirname = dirname(program_name)
+    if isempty(program_dirname)
+        # If we have been given just a program name (not a relative or absolute
+        # path) then we should search `PATH` for it here:
+        pathsep = @static iswindows() ? ';' : ':'
+        path_dirs = split(get(ENV, "PATH", ""), pathsep)
+
+        # On windows we always check the current directory as well
+        @static if iswindows()
+            pushfirst!(path_dirs, 1, pwd())
+        end
+    else
+        # If we've been given a path that has a directory name in it, then we
+        # check to see if that path exists (with all the attendant `program_names`
+        # skullduggery windows requires).  Otherwise, we search the PATH.
+        push!(program_dirname)
+    end
+
+    for path_dir in path_dirs
+        for program_name in program_names
+            program_path = joinpath(path_dir, program_name)
+            # If we find something that matches our name and we can execute
+            if isexecutable(program_path)
+                return abspath(program_path)
```

Comment:
`abspath` doesn't compute the same location as `isexecutable`. Did you mean `realpath` (which would match the binutils `which` program) or to call `abspath` first (which would be odd)?

Reply:
You're right; the `abspath()` was vestigial from an older way of doing things; now that `program_path` is a `joinpath()` output, we never have the case where we need to turn a relative path into an absolute one; this path is always absolute.